federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+// federationSyncPeriod controls how often we check the federation's services to
+// ensure that the correct Kubernetes services (and associated DNS entries) exist.
+// This is only necessary to fudge over failed watches.
+// clusterSyncPeriod controls how often we check the federation's underlying clusters and
+// their Kubernetes services to ensure that matching services created independently of the Federation
+// (e.g. directly via the underlying cluster's API) are correctly accounted for.
+
+// It's an error to call Run() more than once for a given ServiceController
+// object.
+func (sc *ServiceController) Run(workers int, stopCh <-chan struct{}) error {
+	if err := sc.init(); err != nil {
+		return err
+	}
+
+	defer runtime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	go sc.clusterController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	go wait.Until(sc.clusterServiceSyncLoop, clusterSyncPeriod, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down Service Controller")
+	sc.queue.ShutDown()
+	return nil
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncService(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing service: %v", err)
+			}
+		}()
+	}
+}
+
+func (s *ServiceController) init() error {
+//TODO: waiting for DNS lib
+
+	zones, ok := s.dns.Zones()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support zone enumeration, which is required for creating load balancers.")
+	}
+	zoneList, err := zones.List()
+	if err != nil {
+		return fmt.Errorf("failed to get zone from cloud provider, will not be able to create load balancers: %v", err)
+	}
+	s.zones = zoneList
+	return nil
+}
+
+func wantsDNSRecords(service *v1.Service) bool{
+	return true
+}
+
+func (s *ServiceController) processServiceForCluster(key, clusterName string, service, appliedState *v1.Service, client clientset.Clientset) (error, bool) {
+	// Create or Update k8s Service
+	s.ensureClusterServices(service, client, clusterName)
+	if appliedState != nil && !s.needsUpdate(appliedState, service) {
+		glog.Infof("DNS records don't need update for service %s", key)
+		return nil, notRetryable
+	}
+
+	// Note: It is safe to just call EnsureDNSRecords.  But, on some clouds that requires a delete & create,
+	// which may involve service interruption.  Also, we would like user-friendly events.
+
+	// Save the state so we can avoid a write if it doesn't change
+	//TODO: jesse
+	//previousState := api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer) // TODO: quinton: are the DNS names in here?  I think so.
```

Comment:
To answer my own question, yes.  Each service status's loadbalancer has possibly multiple ingresses, each with an IP and hostname.  For Ubernetes services, the ip and hostname of every underlying cluster service's loadbalancer ingresses should be added:

Relevant part of kubernetes API spec.

```
  "status": {
    "loadBalancer": {
      "ingress": [
        {
          "ip": "string",
          "hostname": "string"
        }
      ]
```